packages/http-client/src/csrf-token-middleware.ts

```diff
@@ -0,0 +1,199 @@
+import {
+  createLogger,
+  ErrorWithCause,
+  first,
+  flatten,
+  pickIgnoreCase,
+  pickValueIgnoreCase,
+  removeTrailingSlashes
+} from '@sap-cloud-sdk/util';
+import axios from 'axios';
+import { executeWithMiddleware } from '@sap-cloud-sdk/resilience/internal';
+import {
+  HttpMiddleware,
+  HttpMiddlewareContext,
+  HttpMiddlewareOptions,
+  HttpRequestConfig,
+  HttpRequestConfigBase,
+  HttpRequestConfigWithOrigin,
+  Method
+} from './http-client-types';
+
+const logger = createLogger('csrf-middleware');
+
+/**
+ * Options for middleware to fetch csrf tokens.
+ */
+export interface CsrfMiddlewareOptions {
+  /**
+   * Method used for the token fetching. Default is `head`.
+   */
+  method?: Method;
+  /**
+   * Url used for the token fetching. Default is the resource path without parameters.
+   */
+  url?: string;
+  /**
+   * Middlewares added to the token retrieval request.
+   */
+  middleware?: HttpMiddleware[];
+}
+
+/**
+ * Middleware for fetching a CSRF token. This middleware is added to all request per default.
+ * Use the `fetchCsrfToken` option to disable it.
+ * @param options - Options like URL or method to configure the token fetching.
+ * @returns The middleware for fetching CSRF tokens.
+ */
+export function csrf(options?: CsrfMiddlewareOptions): HttpMiddleware {
+  return (middlewareOptions: HttpMiddlewareOptions) => async requestConfig => {
+    if (noActionNeeded(requestConfig)) {
```

Comment:
[pp] I prefer to test the positive case to avoid confusion through potential double negation. See my comments below, I would find it more clear like this:
```suggestion
    if (!needsCsrfToken(requestConfig)) {
```